pixels-turbo/pixels-worker-vhive/src/main/java/io/pixelsdb/pixels/worker/vhive/PixelsWriterStreamImpl.java

```diff
@@ -0,0 +1,830 @@
+                } catch (IOException e) {
+                    throw new CompletionException("failed to write column vector", e);
+                }
+            });
+            futures[i] = future;
+        }
+        CompletableFuture.allOf(futures).join();
+        curRowGroupDataLength += dataLength.get();
+    }
+
+    /**
+     * Close PixelsWriterStreamImpl, indicating the end of stream.
+     */
+    @Override
+    public void close() {
+        try {
+            if (curRowGroupNumOfRows != 0) {
+                writeRowGroup();
+            }
+            // If the outgoing stream is empty (addRowBatch() and thus writeRowGroup() never called), we artificially send an empty row group here before closing,
+            //  so that the HTTP server can properly move on and close.
+            else if (isFirstRowGroup) {
+                writeRowGroup(); isFirstRowGroup = false;
+            }
+
+            if (!partitioned) {
+                // close the HTTP connection
+                // XXX: Can just set the connection header to "close" on the last row group / each partitioned row group,
+                //  instead of sending a separate close request. -- the PixelsWriterStreamImpl does not know which row group is the last. Not possible.
+                if (!partitioned && uri == null) {
+                    uri = URI.create(fileNameToUri(fileName));
+                }
+                Request req = httpClient.preparePost(partitioned ? fileNameToUri(fileNames.get(currHashValue)) : uri.toString())
+                        .addHeader(CONTENT_TYPE, "application/x-protobuf")
+                        .addHeader(CONTENT_LENGTH, 0)
+                        .addHeader(CONNECTION, CLOSE)
+                        .build();
+                try {
+                    outstandingHTTPRequestSemaphore.acquire();
+                    Response response = httpClient.executeRequest(req).get();
+                    if (response.getStatusCode() != 200) {
+                        throw new IOException("Failed to send close request to server. Is the server already closed? HTTP status code: " + response.getStatusCode());
+                    }
+                } catch (Throwable e) {
+                    // log to warn that remote has already closed, but not throwing out the exception
+                    logger.warn(e.getMessage());
+                    e.printStackTrace();
+                }
+            }
+
+            for (ColumnWriter cw : columnWriters) {
+                cw.close();
+            }
+            columnWriterService.shutdown();
+            columnWriterService.shutdownNow();
+
+            if (byteBuf.refCnt() > 0) {
+                byteBuf.release();
+            }
+        } catch (IOException e) {
+            logger.error(e.getMessage());
+            e.printStackTrace();
+        }
+    }
+
+    // XXX: In demo version, I have modified this method to be non-private, because StreamWorkerCommon::passSchemaToNextLevel() needs it.
```

Comment:
Please fix this todo and revise the comment. Why do you expose writeRowGroup to the other classes in this package?

Reply:
In `StreamWorkerCommon`, we utilize the `writeRowGroup()` to send the schema (in the packet's `StreamHeader`) to the next level worker. Not exposing this method could necessitate copying this code to `StreamWorkerCommon`, although there are other solutions.
Resolved in later commits.